uber/site_sections/budget.py

```diff
@@ -139,44 +142,89 @@ def generate_promo_codes(
+            'uses_allowed': uses_allowed
+        }
+
         if cherrypy.request.method == 'POST':
             codes = None
             if is_single_promo_code:
-                if params.get('code', '').strip():
-                    codes = [params['code']]
-                else:
-                    count = 1
+                count = 1
+                if code:
+                    codes = [code]
 
             if use_words and not codes and \
                     not any(s for (_, s) in words.items()):
-                raise HTTPRedirect('generate_promo_codes?message='
-                    'Please add some promo code words!')
+                result['message'] = 'Please add some promo code words!'
+                return result
 
             if not codes:
                 if use_words:
                     codes = PromoCode.generate_word_code(count)
                 else:
-                    length = int(params.get('length', 12))
-                    segment_length = int(params.get('segment_length', 3))
+                    try:
+                        length = int(length)
+                    except:
+                        length = 9
+                    try:
+                        segment_length = int(segment_length)
+                    except:
+                        segment_length = 3
                     codes = PromoCode.generate_random_code(
                         count, length, segment_length)
 
             promo_codes = []
             for code in codes:
-                params['code'] = code
-                promo_codes.append(PromoCode().apply(params, restricted=False))
+                promo_codes.append(PromoCode().apply({
+                        'code': code,
+                        'expiration_date': expiration_date,
+                        'discount_type': discount_type,
+                        'discount': discount,
+                        'uses_allowed': uses_allowed,
+                    }, restricted=False))
 
             message = check_all(promo_codes)
             if message:
-                raise HTTPRedirect('generate_promo_codes?message={}', message)
+                result['message'] = message
+                return result
 
             result['promo_codes'] = session.bulk_insert(promo_codes)
-            if len(result['promo_codes']) != count:
+            generated_count = len(result['promo_codes'])
+            if generated_count <= 0:
+                result['message'] = "Could not generate any of the " \
+                    "requested promo codes. Perhaps they've all been taken " \
+                    "already?"
+                return result
+
+            if generated_count != count:
                 result['message'] = 'Some of the requested promo codes ' \
                     'could not be generated'
 
         if 'export' in params:
             return self.export_promo_codes(codes=result['promo_codes'])
+
+        result['is_single_promo_code'] = 1
```

Comment:
Ooh, I see. I think there is!

Instead of painstakingly building a custom dictionary, you can get a lot of (though not all) these defaults from the model itself by passing an empty/new PromoCode() object to the template. Then in the template you'd use, e.g., `{{ promo_code.count }}` instead of `{{ count }}`. This will save you from having to instantiate `code`, `discount` (though maybe not... looks like you have a different default here than on the model), `discount_type`, `expiration_date`, and `uses_allowed` (same as the discount, though, you've set it to something different here). Using the DB defaults as our form defaults is a super common pattern in the code, as you've probably noticed by now. You can ALSO return a custom dict with other defaults, of course. I'd recommend using the model if at all possible, since it also prevents weird mismatches between what the default is in this function and what it is in the database.